mirrord/agent/src/steal/connection.rs

```diff
@@ -347,10 +369,35 @@ impl TcpConnectionStealer {
 
     /// Removes the ([`ReadHalf`], [`WriteHalf`]) pair of streams, disconnecting the remote
     /// connection.
+    /// Also remove connection from connection mappings and free the connection index.
+    /// This method does not remove from client_connections so that it can be called while
     #[tracing::instrument(level = "trace", skip(self))]
-    fn connection_unsubscribe(&mut self, connection_id: ConnectionId) {
+    fn remove_connection(&mut self, connection_id: ConnectionId) -> Option<ClientId> {
         self.write_streams.remove(&connection_id);
         self.read_streams.remove(&connection_id);
+        self.index_allocator.free_index(connection_id);
+        self.connection_client.remove(&connection_id)
+    }
+
+    /// Close the connection, remove the id from all maps and free the id.
+    #[tracing::instrument(level = "trace", skip(self))]
+    fn connection_unsubscribe(&mut self, connection_id: ConnectionId) {
+        if let Some(client_id) = self.remove_connection(connection_id) {
+            // Remove the connection from the set of the connections that belong to its client.
+            self.client_connections
+                .entry(client_id)
+                .and_modify(|connections| {
+                    connections.remove(&connection_id);
+                });
+            // If we removed the last connection of this client, remove client from map.
+            if self
```

Comment:
Can't you remove the connection within the `and_modify` block above? If not, maybe just check that the connections are empty within `and_modify` and toggle some flag, then use that to remove the client, avoiding getting the client connection three times?

Reply:
Can't remove it inside `and_modify`, because `self.client_connections.remove` borrows `self` as mutable, and `self` is already borrowed in the `entry(..).and_modify(..)` call.

Added the flag, doing one less hashmap lookup, thanks! :)